File: Python/clbg/fasta/fasta.py
import sys

def main():
    N = int(sys.argv[1])

    out = sys.stdout.write

    alu = (
        "GGCCGGGCGCGGTGGCTCACGCCTGTAATCCCAGCACTTTGG"
        "GAGGCCGAGGCGGGCGGATCACCTGAGGTCAGGAGTTCGAGA"
        "CCAGCCTGGCCAACATGGTGAAACCCCGTCTCTACTAAAAAT"
        "ACAAAAATTAGCCGGGCGTGGTGGCGCGCGCCTGTAATCCCA"
        "GCTACTCGGGAGGCTGAGGCAGGAGAATCGCTTGAACCCGGG"
        "AGGCGGAGGTTGCAGTGAGCCGAGATCGCGCCACTGCACTCC"
        "AGCCTGGGCGACAGAGCGAGACTCCGTCTCAAAAA"
    )
    alu_len = len(alu)

    def write_wrapped_repeat(length):
        idx = 0
        remaining = length
        while remaining > 0:
            line_len = 60 if remaining >= 60 else remaining
            if idx + line_len <= alu_len:
                out(alu[idx:idx + line_len] + "\n")
                idx += line_len
                if idx == alu_len:
                    idx = 0
            else:
                part1 = alu[idx:]
                part2_len = line_len - len(part1)
                out(part1 + alu[:part2_len] + "\n")
                idx = part2_len
            remaining -= line_len

    IM = 139968
    IA = 3877
    IC = 29573
    seed = 42

    def make_table(items):
        # items: list of (char, probability)
        table = []
        acc = 0.0
        for ch, p in items:
            acc += p
            table.append((acc, ch))
        return table

    iub = make_table([
        ('a', 0.27), ('c', 0.12), ('g', 0.12), ('t', 0.27),
        ('B', 0.02), ('D', 0.02), ('H', 0.02), ('K', 0.02),
        ('M', 0.02), ('N', 0.02), ('R', 0.02), ('S', 0.02),
        ('V', 0.02), ('W', 0.02), ('Y', 0.02),
    ])

    homo = make_table([
        ('a', 0.3029549426680),
        ('c', 0.1979883004921),
        ('g', 0.1975473066391),
        ('t', 0.3015094502008),
    ])

    def write_random(length, table):
        nonlocal seed
        remaining = length
        vals = [x[0] * IM for x in table]
        chars = [x[1] for x in table]
        tlen = len(table)

        while remaining > 0:
            line_len = 60 if remaining >= 60 else remaining
            line_chars = []
            for _ in range(line_len):
                seed = (seed * IA + IC) % IM
                r = seed
                # binary search would be fine, but tiny tables => linear is fastest here
                for i in range(tlen):
                    if r < vals[i]:
                        line_chars.append(chars[i])
                        break
            out(''.join(line_chars) + "\n")
            remaining -= line_len

    out(">ONE Homo sapiens alu\n")
    write_wrapped_repeat(N * 2)

    out(">TWO IUB ambiguity codes\n")
    write_random(N * 3, iub)

    out(">THREE Homo sapiens frequency\n")
    write_random(N * 5, homo)

if __name__ == "__main__":
    main()

File: TypeScript/clbg/fasta/fasta.js
const N = parseInt(process.argv[2] || "0", 10);
const IM = 139968;
const IA = 3877;
const IC = 29573;
let seed = 42;
function genRandom() {
    seed = (seed * IA + IC) % IM;
    return seed / IM;
}
function makeCumulative(table) {
    let sum = 0;
    return table.map(([c, p]) => {
        sum += p;
        return [c, sum];
    });
}
const iub = makeCumulative([
    ["a", 0.27],
    ["c", 0.12],
    ["g", 0.12],
    ["t", 0.27],
    ["B", 0.02],
    ["D", 0.02],
    ["H", 0.02],
    ["K", 0.02],
    ["M", 0.02],
    ["N", 0.02],
    ["R", 0.02],
    ["S", 0.02],
    ["V", 0.02],
    ["W", 0.02],
    ["Y", 0.02],
]);
const homo = makeCumulative([
    ["a", 0.3029549426680],
    ["c", 0.1979883004921],
    ["g", 0.1975473066391],
    ["t", 0.3015094502008],
]);
const alu = "GGCCGGGCGCGGTGGCTCACGCCTGTAATCCCAGCACTTTGG" +
    "GAGGCCGAGGCGGGCGGATCACCTGAGGTCAGGAGTTCGAGA" +
    "CCAGCCTGGCCAACATGGTGAAACCCCGTCTCTACTAAAAAT" +
    "ACAAAAATTAGCCGGGCGTGGTGGCGCGCGCCTGTAATCCCA" +
    "GCTACTCGGGAGGCTGAGGCAGGAGAATCGCTTGAACCCGGG" +
    "AGGCGGAGGTTGCAGTGAGCCGAGATCGCGCCACTGCACTCC" +
    "AGCCTGGGCGACAGAGCGAGACTCCGTCTCAAAAA";
let out = [];
function emitHeader(s) {
    out.push(s);
}
function writeRepeat(seq, length) {
    const slen = seq.length;
    let idx = 0;
    let line = "";
    for (let i = 0; i < length; i++) {
        line += seq[idx];
        idx++;
        if (idx === slen)
            idx = 0;
        if (line.length === 60) {
            out.push(line);
            line = "";
        }
    }
    if (line.length)
        out.push(line);
}
function writeRandom(cum, length) {
    let line = "";
    for (let i = 0; i < length; i++) {
        const r = genRandom();
        let lo = 0, hi = cum.length - 1;
        while (lo < hi) {
            const mid = (lo + hi) >> 1;
            if (r < cum[mid][1])
                hi = mid;
            else
                lo = mid + 1;
        }
        line += cum[lo][0];
        if (line.length === 60) {
            out.push(line);
            line = "";
        }
    }
    if (line.length)
        out.push(line);
}
emitHeader(">ONE Homo sapiens alu");
writeRepeat(alu, 2 * N);
emitHeader(">TWO IUB ambiguity codes");
writeRandom(iub, 3 * N);
emitHeader(">THREE Homo sapiens frequency");
writeRandom(homo, 5 * N);
process.stdout.write(out.join("\n") + "\n");
